datacube/index/abstract.py

```diff
@@ -922,28 +922,49 @@ def find_less_mature(self, ds: Dataset, delta: int = 500) -> Iterable[Dataset]:
         """
         less_mature = []
         assert delta >= 0
+
+        def check_maturity_information(dataset, props):
+            # check that the dataset metadata includes all maturity-related properties
+            # passing in the required props to enable greater extensibility should it be needed
+            for prop in props:
+                if hasattr(dataset.metadata, prop) and (getattr(dataset.metadata, prop) is not None):
+                    return
+                raise ValueError(
+                    f"Dataset {dataset.id} is missing property {prop} required for maturity check"
+                )
+
+        check_maturity_information(ds, ["region_code", "time", "dataset_maturity"])
+
```

Comment:
I'd expect a sanity check on `dataset_maturity` happens first. If the dataset doesn't have this item, there is no need to go through archiving process. Then it doesn't matter if dataset has any required keys.

Reply:
or it implies the code shouldn't get here if some dataset doesn't have these keys?

Reply:
Does it matter which property is checked first if all of them ultimately need to be present? This logic gets called before anything else gets done in `archive_less_mature`, so I'm not sure how much earlier a `dataset_maturity` sanity check could be done anyway.

Reply:
I guess I got some misconception about the code, which requires some clarification
- will/should the code be called if the product doesn't need `archive_less_mature`, which certainly doesn't have `dataset_maturity` in the metadata (might/not miss `region_code`)
- if the code get called when the product doesn't need `archive_less_mature`, what is the expected behaviour? throw the error and expect some other code to deal with it, if so, where? or give a warning and proceed into usual indexing.

In my understanding, whether or not having `dataset_maturity` is a criterion to tell if `archive_less_mature` is needed. So it's a safe pin for sanity check. In that case, if having `region_code` is tolerable.

Reply:
`archive_less_mature` is only ever called if the flag is provided by the user. Ideally, if a product does not have dataset maturity, the `--archive-less-mature` flag would not be set to begin with.
If the flag is set, then line 936 would be as far as the code would get, because `check_maturity_information` would throw and error. So none of the other logic would be called, whether or not the dataset is missing `region_code`. I suppose you could conduct the check slightly earlier, but I don't think it would make much of a difference in terms of what actually gets executed.
Based on how the check was originally in `odc-tools`, I assumed the expected behaviour would be for the check to throw an error. But it could equally just raise a warning and skip the dataset. It would then either be handled by the user via the cli, or managed by pipeline code if that's where it's being called.

Reply:
OK! Then I'd settle with throwing the error here and making `odc-tools` catch and handle. Thanks a lot!